metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java

```diff
@@ -309,8 +309,10 @@ public void activate() {
         long nowNs = time.nanoseconds();
         for (BrokerRegistration registration : brokerRegistrations.values()) {
             heartbeatManager.register(registration.id(), registration.fenced());
-            heartbeatManager.tracker().updateContactTime(
-                new BrokerIdAndEpoch(registration.id(), registration.epoch()), nowNs);
+            if (!registration.fenced()) {
+                heartbeatManager.tracker().updateContactTime(
+                    new BrokerIdAndEpoch(registration.id(), registration.epoch()), nowNs);
```

Comment:
Interesting. This is not a new issue but this means that a cluster with frequent controller failovers, more often than the heartbeat timeout, will be unable to fence brokers that have not sent a heartbeat.

Have you considered persisting some of the session state? What is the scalability impact of persisting session state? For example, my understanding is that ZooKeeper persists session state. That is how they implement ephemeral nodes.

Reply:
In general there can be a lot of session state. Maybe we'll persist it in the future, but I don't think something as minor as this would be the motivation.

One thing we could do is have separate states for never unfenced (initial startup) and post-controlled-shutdown. That would let us have some better metrics. But it's not worth it for this PR.